File: BEmuJava/bemu/src/main/java/com/bloomberglp/blpapi/HistoricRequestElementStringArray.java
//------------------------------------------------------------------------------
// <copyright project="BEmu_maven" file="/BEmu_maven/bemu/src/main/java/com/bloomberglp/blpapi/HistoricRequestElementStringArray.java" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

package com.bloomberglp.blpapi;

import java.util.ArrayList;
import java.util.List;

import com.bloomberglp.blpapi.Name;
import com.bloomberglp.blpapi.Element;

public class HistoricRequestElementStringArray extends Element
{
    private final String _elementName;
    private final List<String> _values;

    HistoricRequestElementStringArray(String elementName)
    {
        this._elementName = elementName;
        this._values = new ArrayList<String>();
    }
    
    public Name name() throws Exception
    {
    	return new Name(this._elementName);
    }
    
    public int numValues()
    {
    	return this._values.size();
    }
    
    public int numElements()
    {
    	return 0;
    }
    
    public boolean isComplexType()
    {
    	return false;
    }
    
    public boolean isArray()
    {
    	return true;
    }
    
    public boolean isNull()
    {
    	return false;
    }
    
    void appendValue(String value)
    {
    	this._values.add(value);
    }
    
    List<String> values()
    {
    	return this._values;
    }
    
    protected StringBuilder prettyPrint(int tabIndent)
    {
        String tabs = com.bloomberglp.blpapi.IndentType.Indent(tabIndent);
        StringBuilder result = new StringBuilder();

        result.append(String.format("%s%s[] = {%s", tabs, this._elementName, System.getProperty("line.separator")));

        StringBuilder secs = new StringBuilder();
        for(int i = 0; i < this._values.size(); i++)
        {
        	secs.append(this._values.get(i));
        	if(i < this._values.size() - 1)
        		secs.append(",");
        }
        result.append(String.format("%s%s%s%s", tabs, com.bloomberglp.blpapi.IndentType.TAB, secs.toString(), System.getProperty("line.separator")));
        
        result.append(String.format("%s}%s", tabs, System.getProperty("line.separator")));

        return result;
    }
}

File: BEmu_cpp/src/HistoricalDataRequest/HistoricRequestElementStringArray.cpp
//------------------------------------------------------------------------------
// <copyright project="BEmu_cpp" file="src/HistoricalDataRequest/HistoricRequestElementStringArray.cpp" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

#include "HistoricalDataRequest/HistoricRequestElementStringArray.h"
#include "BloombergTypes/Name.h"
#include "Types/IndentType.h"

namespace BEmu
{
	namespace HistoricalDataRequest
	{
		HistoricRequestElementStringArray::HistoricRequestElementStringArray(const std::string& elementName)
		{
			this->_elementName = elementName;
		}

		HistoricRequestElementStringArray::~HistoricRequestElementStringArray()
		{
			this->_values.clear();
		}


		Name HistoricRequestElementStringArray::name() const
		{
			Name result(this->_elementName.c_str());
			return result;
		}

		size_t HistoricRequestElementStringArray::numValues() const
		{
			return this->_values.size();
		}

		size_t HistoricRequestElementStringArray::numElements() const
		{
			return 0;
		}

		
		bool HistoricRequestElementStringArray::isNull() const
		{
			return false;
		}

		bool HistoricRequestElementStringArray::isArray() const
		{
			return true;
		}

		bool HistoricRequestElementStringArray::isComplexType() const
		{
			return false;
		}
		
		std::vector<std::string> HistoricRequestElementStringArray::values() const
		{
			return this->_values;
		}

		void HistoricRequestElementStringArray::appendValue(const char* value)
		{
			this->_values.push_back(value);
		}


		std::ostream& HistoricRequestElementStringArray::print(std::ostream& stream, int level, int spacesPerLevel) const
		{
			std::string tabs(IndentType::Indent(level, spacesPerLevel));

			stream << tabs << this->_elementName << "[] = {" << std::endl;
			stream << tabs << IndentType::Indent(1, spacesPerLevel);

			int size = this->_values.size();
			int count = 0;
			for(std::vector<std::string>::const_iterator iter = this->_values.begin(); iter != this->_values.end(); ++iter)
			{
				std::string str = *iter;

				stream << "\"" << str << "\"";
				if(count < size - 1)
					stream << ", ";

				count++;
			}

			stream << std::endl;			
			stream << tabs << '}' << std::endl;

			return stream;
		}

	}
}